### PR TITLE
AsyncProducer: Limit the number of retries & AsyncProducer: make sure threads are running before shutdown

### LINKED ISSUE #707: AsyncProducer will enter to an infinite loop when the internal `sync_producer` will keep raising `Kafka::BufferOverflow`
* Version of Ruby: 2.4.2
* Version of Kafka: 2.0
* Version of ruby-kafka: 0.7.5

###### Steps to reproduce

In the `async_producer`, private method `produce` will enter to an infinite loop when the internal `sync_producer` will keep raising `Kafka::BufferOverflow`.

```
      def produce(*args)
        @producer.produce(*args)
      rescue BufferOverflow
        deliver_messages
        retry
      end
```

###### Please write an example that reproduces the problem you're describing.

Send at least one message above the `producer.max_buffer_size` (default 1000) and turn off all the brokers. 

###### Expected outcome

There should be a limit number of retries for `producer.produce` call.

###### Actual outcome

Endless loop.

### PR DESCRIPTION
**Pull request by dhruvparanjape**
* fixes https://github.com/zendesk/ruby-kafka/issues/707 which can happen when no kafka cluster is running example when running tests locally.
* if a worker thread unexpectedly dies with unhandled messages in its queue and then we try to call `shutdown`, those messages won't be handled unless `produce` is called in the future (in which case the worker thread will read the `shutdown` message enqueued here and then exit again). With this change we will restart the worker thread so it can receive the `shutdown` message.

Both changes are backwards compatible.

Tested event production on staging with `cw-finance` and `cw-buyer` producing the currency created event.

### CODE REVIEW

File: lib/kafka/async_producer.rb [hunk 1]
```diff
@@ -179,10 +196,12 @@ def run
     end
 
     class Worker
-      def initialize(queue:, producer:, delivery_threshold:, instrumenter:, logger:)
+      def initialize(queue:, producer:, delivery_threshold:, max_retries: -1, retry_backoff: 0, instrumenter:, logger:)
```
Comment (tdyas) [inline lib/kafka/async_producer.rb]: Instead of `-1` as the magic value, would `nil` be a better choice since it represents absence of a value?

Comment (dhruvparanjape) [reply lib/kafka/async_producer.rb]: makes sense 👍  changed.

**Review by ahmad-hussain-catawiki — APPROVED** [review]
Seems good

**Comment by sarcilav** [comment]
Too late, but my 2 cents here, the two commits were backports, it was probably better to keep as they were, in case you need to keep backporting stuff

**Comment by dhruvparanjape** [comment]
>Too late, but my 2 cents here, the two commits were backports, it was probably better to keep as they were, in case you need to keep backporting stuff

Hey 👋 

yup that was a trade off but we don't expect to backport a lot of changes going forwards (hopefully for the next coming half year) before replacing the client.

**Pull request merged**